src/test_typing_extensions.py

```diff
@@ -850,6 +857,49 @@ def d(): pass
             isinstance(cell.cell_contents, deprecated) for cell in d.__closure__
         ))
 
+@deprecated("depr")
+def func():
+    pass
+
+@deprecated("depr")
+async def coro():
+    pass
+
+class Cls:
+    @deprecated("depr")
+    def func(self):
+        pass
+
+    @deprecated("depr")
+    async def coro(self):
+        pass
+
+class DeprecatedCoroTests(BaseTestCase):
+    @skipUnless(TYPING_3_12_ONLY or TYPING_3_13_0_RC, "inspect.iscoroutinefunction works differently on Python < 3.12")
+    def test_inspect_py313(self):
+        self.assertFalse(inspect.iscoroutinefunction(func))
+        self.assertFalse(inspect.iscoroutinefunction(Cls.func))
+        self.assertTrue(inspect.iscoroutinefunction(coro))
+        self.assertTrue(inspect.iscoroutinefunction(Cls.coro))
+
+        self.assertFalse(asyncio.coroutines.iscoroutinefunction(func))
+        self.assertFalse(asyncio.coroutines.iscoroutinefunction(Cls.func))
+        self.assertTrue(asyncio.coroutines.iscoroutinefunction(coro))
+        self.assertTrue(asyncio.coroutines.iscoroutinefunction(Cls.coro))
+
+    @skipIf(TYPING_3_12_0, "inspect.iscoroutinefunction works differently on Python 3.12+")
+    def test_inspect_py311(self):
+        # This doesn't work in Python < 3.12.
+        # self.assertTrue(inspect.iscoroutinefunction(func))
+        # self.assertTrue(inspect.iscoroutinefunction(Cls.func))
+        self.assertFalse(inspect.iscoroutinefunction(coro))
+        self.assertFalse(inspect.iscoroutinefunction(Cls.coro))
+
+        self.assertFalse(asyncio.coroutines.iscoroutinefunction(func))
+        self.assertFalse(asyncio.coroutines.iscoroutinefunction(Cls.func))
+        self.assertTrue(asyncio.coroutines.iscoroutinefunction(coro))
+        self.assertTrue(asyncio.coroutines.iscoroutinefunction(Cls.coro))
```

Comment:
The `asyncio.coroutines.iscoroutine()` assertions are the same on all Python versions. Why don't we separate them out into their own test, so that it's obvious to the reader which behaviours vary according to Python version?

I'm also still not sure of the value of making any assertions about what `inspect.iscoroutinefunction` does on Python 3.11 and lower, since it seems we can't get desirable behaviour on those versions.

```suggestion
    def test_asyncio_iscoroutinefunction(self):
        self.assertFalse(asyncio.coroutines.iscoroutinefunction(func))
        self.assertFalse(asyncio.coroutines.iscoroutinefunction(Cls.func))
        self.assertTrue(asyncio.coroutines.iscoroutinefunction(coro))
        self.assertTrue(asyncio.coroutines.iscoroutinefunction(Cls.coro))

    @skipUnless(TYPING_3_12_ONLY or TYPING_3_13_0_RC, "inspect.iscoroutinefunction works differently on Python < 3.12")
    def test_inspect_iscoroutinefunction(self):
        self.assertFalse(inspect.iscoroutinefunction(func))
        self.assertFalse(inspect.iscoroutinefunction(Cls.func))
        self.assertTrue(inspect.iscoroutinefunction(coro))
        self.assertTrue(inspect.iscoroutinefunction(Cls.coro))